Corrected the graph export set expansion rules that are described at …

szoupanos opened the pull request:
…issue 1102

https://github.com/aiidateam/aiida_core/issues/1102

There were problems when overriding the default behaviour of the rules. Moreover
they were moved to separate methods to resuce the complexity of the export method.

Various tests were added:
- Export tests that are based on a real workchain were added to ensure that a complex
  and real AiiDA graph based on workchains is exported correctly.
- Export tests based on a synthetic graph were added that explicitly check every
  flag that changes the default behaviour of the export set expansion rules.

ltalirz commented:
@danieleongari has now done a first test

Here is what we noticed:

 * printing out the selected rules (very helpful!) made us realize that the current implementation of the `-C` option is very counterintuitive: while all other options turn the corresponding rules *on*,  `-C` turns the rule *off*
```
  -C, --create-reversed           Follow reverse CREATE links (recursively)
                                  when calculating the node set to export.
                                  [default: True]
```
I think the only proper way of resolving this is to have both `--create-reversed` and `--no-create-reversed` options (same for all other options).

 * When turning on all possible rules (`-RIX`), the export got stuck (we canceled after a few minutes). 
   I wonder what it was looking for - would it have tried to take the whole connected subset of the graph? (The DB contains 194125 nodes in total)

 * The option `-RX`now seems to produce what we were looking for originally :-)

 * Printing uuids instead of PKs for `verdi graph generate` is a good idea - however
   * it is important to use a shortened version of the uuid (see [graph.pdf](https://github.com/aiidateam/aiida_core/files/3000336/113.pdf); to be compared to the [original graph](https://github.com/aiidateam/aiida_core/files/2946116/103481.pdf) )
   * it should be possible to switch back to PKs if desired

ltalirz commented:
Concerning which subset of the uuid to show: git uses 7 characters for the commit hash; we could use 8 characters (the first part of the uuid).

Subsets of uuids are already supported in 0.12, e.g. you can do
`verdi  node show -u ea70cc12`

szoupanos commented:
Thanks a lot Leopold for the testing and the feedback!

Due to time constrains (and for keeping things simple), I would keep this PR to the absolutely necessary to make the export rules work correctly (the UUID at the graph generate were left there by accident because they were useful when debugging - but if we all like it, I can keep a shorter UUID version).

The discussion on overriding the arguments can be added at another ticket and the implementation at another PR.

ltalirz commented:
> Due to time constrains (and for keeping things simple), I would keep this PR to the absolutely necessary to make the export rules work correctly 

Sounds good to me. Let's do both the uuid stuff and the changes to options in a separate PR (please remove the uuid changes).

I will review the code changes this evening.

ltalirz commented:
However, please keep printing the rules that are selected in this PR (perhaps add a print "Rules on following links:" line before)

szoupanos commented:
The addition of the UUIDs at the graph will be addressed at issue #2636

szoupanos commented:
Further discussion on the export set expansion rules can be done at #2637

szoupanos commented:
For me it is OK.

- Switched back to pks for the AiiDA graph
- Created the necessary tickets for the AiiDA graph (with UUIDS) and the default behaviour of the export rules

ltalirz commented on aiida/backends/tests/export_and_import.py:
```diff
@@ -1481,6 +1483,175 @@ def get_all_node_links(self):
+        (INPUT, CREATE, RETURN and CALL)
+        """
+
+        test_case = None
```
should these be class variables?
if they are just instance variables, you can remove these two lines

szoupanos replied:
I prefer to keep them for clarity - I don't like the approach of using undefined variables even if Python allows it - I find it dirty

szoupanos replied:
But I will document them though

ltalirz commented on aiida/backends/tests/export_and_import.py:
```diff
@@ -1481,6 +1483,175 @@ def get_all_node_links(self):
                   edge_project=['label', 'type'], output_of='input')
         return qb.all()
 
+    class ComplexGraph:
```
this file already has >1k lines
could you please move the ComplexGraph out into a separate file?

szoupanos replied:
I propose this to be done under another PR. And mainly to 1.0
There is already one created by Casper (#2702)

ltalirz commented on aiida/backends/tests/export_and_import.py:
```diff
@@ -1481,6 +1483,175 @@ def get_all_node_links(self):
+            wc2 = WorkCalculation().store()
+
+            pw1 = JobCalculation()
+            pw1.set_computer(self.test_case.computer)
```
you use the test_case just to get the computer, correct?
you use the `test_case` just for the computer, correct?
can you just pass the `computer` to the constructor instead?

this is more explicit, while passing the test case creates a loop where the test case creates the ComplexGraph, which again has a reference to the test case, ...

szoupanos replied:
Good point, I will do that.
I passed the test_case since I wanted to be more generic.
But it is cleaner what you propose.

ltalirz commented:
I have to stop the review now, will continue later

szoupanos commented:
I addressed some of the comments. I would propose to merge this and apply the same changes to 1.0 (since it is important to have correct export rules) and leave the further refactoring to be done under another ticket (since there is one - and Casper has already started working on it).

szoupanos commented:
@ltalirz @sphuber 
Can this be merged please?

ltalirz commented:
I'll finish the review tonight

On Thu, Apr 4, 2019, 20:11 szoupanos <notifications@github.com> wrote:

> @ltalirz <https://github.com/ltalirz> @sphuber
> <https://github.com/sphuber>
> Can this be merged please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aiidateam/aiida_core/pull/2632#issuecomment-480005679>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABASPSizZdRoMYHtXWeP_iGnKpM-KkD_ks5vdkBRgaJpZM4cELG9>
> .
>

szoupanos commented:
Thanks a lot Leo!
Yes, with these flags on, according to the following rules the full sub-graph that is connected to the node should be returned.
https://github.com/aiidateam/aiida_core/issues/1102#issuecomment-383090834

Regarding the speed of the graph exploration algorithm, I think that it is a simple graph traversal algorithm
https://en.wikipedia.org/wiki/Graph_traversal

If you find a case that you think that it is particularly slow, we can have a look at it (in case there is a bug etc).
